[AArch64][CodeGen] Optimize security cookie check with New Fixup Pass

omjavaid opened the pull request:
This patch adds the AArch64WinFixupBufferSecurityCheckPass to optimize the handling of buffer security checks on AArch64 Windows targets.

The pass selectively replaces __security_check_cookie calls with inline comparisons, transferring control to the runtime library only on failure.

A similar implementation for X86 Windows target was implemented by #95904

Fixes #117618

omjavaid committed:
Optimize security cookie check with New Fixup Pass

This patch adds the AArch64WinFixupBufferSecurityCheckPass to optimize
the handling of buffer security checks on AArch64 Windows targets.

The pass selectively replaces __security_check_cookie calls with inline
comparisons, transferring control to the runtime library only on failure.

A similar implementation for X86 Windows target was implemented by #95904

omjavaid committed:
Fix arsenm's review comments

efriedma-quic commented on llvm/lib/Target/AArch64/AArch64WinFixupBufferSecurityCheck.cpp:
```diff
@@ -0,0 +1,275 @@
+  MachineBasicBlock::reverse_iterator DEnd = CurMBB->rend();
+
+  // Reverse iterate from CheckCall to find LOAD_STACK_GUARD
+  for (; DIt != DEnd; ++DIt) {
```
Is there some reason to expect there's a LOAD_STACK_GUARD instruction in the same basic block as the __security_check_cookie?  It works out that way in your testcase, but generally I'd expect it to be in the entry block.

omjavaid replied:
Yes you are right. I will fix this in follow up.

omjavaid replied:
Fixed.

efriedma-quic commented on llvm/lib/Target/AArch64/AArch64WinFixupBufferSecurityCheck.cpp:
```diff
@@ -0,0 +1,275 @@
+  MachineBasicBlock::reverse_iterator DIt(CheckCall);
+
+  // Move forward to find the stack adjustment after the call
+  // to __security_check_cookie
```
I'm very skeptical that you'll consistently find all these instructions scheduled exactly the way this code assumes.

omjavaid replied:
IMO Its quite predictable as long as the call to __security_check_cookie expects a single argument and something doesnt change on the MSVC side. But I may be wrong can you kindly explain about how you suspect that this instruction sequence may not always be the same. I can add extra check to make sure that we find this sequence.

I have tried a bit of analysis on different type of function call and always found similar set of instructions like the ones given below:
  Instruction: ADJCALLSTACKDOWN 0, 0, implicit-def dead $sp, implicit $sp
  Instruction: %1:gpr64 = LDRXui %stack.0.StackGuardSlot, 0 :: (volatile load (s64) from %stack.0.StackGuardSlot)
  Instruction: $x0 = COPY %1:gpr64
  Instruction: BL @__security_check_cookie, <regmask $fp $lr $wzr $wzr_hi $xzr $b8 $b9 $b10 $b11 $b12 $b13 $b14 $b15 $d8 $d9 $d10 $d11 $d12 $d13 $d14 $d15 $h8 $h9 $h10 $h11 $h12 $h13 $h14 $h15 $s8 $s9 $s10 $s11 and 92 more...>, implicit-def dead $lr, implicit $sp, implicit $x0, implicit-def $sp
  Instruction: ADJCALLSTACKUP 0, 0, implicit-def dead $sp, implicit $sp

efriedma-quic replied:
I think we usually don't reschedule the code because ADJCALLSTACKDOWN exists at this point in the pipeline, and acts as a scheduling boundary.  Which might work out today, but isn't reliable; it could easily change in the future.

I'm also not sure off the top of my head if we can consistently load the stack guard in a single instruction.

It's probably fine if you abort the transform on an unexpected sequence.

omjavaid replied:
Thanks. I have fixed this to skip transform if sequence doesnt match.

efriedma-quic commented on llvm/test/CodeGen/AArch64/O3-pipeline.ll:
```diff
@@ -159,14 +159,16 @@
 ; CHECK-NEXT:       Remove dead machine instructions
 ; CHECK-NEXT:       AArch64 MI Peephole Optimization pass
 ; CHECK-NEXT:       AArch64 Dead register definitions
+; CHECK-NEXT:       AArch64 Windows Fixup Buffer Security Check
```
Can you preserve the MachineDominator Tree in the new pass?

omjavaid committed:
Fix review comments by Eli

- Fix LOAD_STACK_GUARD detection logic
- Make transform run when security check sequence matched
- Preserve dominator tree

omjavaid commented on llvm/lib/Target/AArch64/AArch64WinFixupBufferSecurityCheck.cpp:
```diff
@@ -0,0 +1,275 @@
+    MachineFunction &MF) {
+  for (auto &MBB : MF) {
+    for (auto &MI : MBB) {
+      if (MI.getOpcode() == AArch64::BL && MI.getNumExplicitOperands() == 1) {
```
Fixed.

omjavaid commented on llvm/lib/Target/AArch64/AArch64WinFixupBufferSecurityCheck.cpp:
```diff
@@ -0,0 +1,275 @@
+          auto Callee = dyn_cast<Function>(MO.getGlobal());
+          if (Callee && Callee->getName() == "__security_check_cookie") {
+            return std::make_pair(&MBB, &MI);
+            break;
```
Fixed.

omjavaid commented on llvm/lib/Target/AArch64/AArch64WinFixupBufferSecurityCheck.cpp:
```diff
@@ -0,0 +1,275 @@
+  void SplitBasicBlock(MachineBasicBlock *CurMBB, MachineBasicBlock *NewRetMBB,
+                       MachineBasicBlock::iterator SplitIt);
+
+  void FinishBlock(MachineBasicBlock *MBB);
```
Fixed.

omjavaid commented on llvm/lib/Target/AArch64/AArch64WinFixupBufferSecurityCheck.cpp:
```diff
@@ -0,0 +1,275 @@
+//===- AArch64WinFixupBufferSecurityCheck.cpp Fix Buffer Security Check Call
+//-===//
```
Fixed.

omjavaid commented:
@efriedma-quic I have addressed your earlier review comments. Kindly let me know if you have any further comments?

efriedma-quic commented on llvm/lib/Target/AArch64/AArch64WinFixupBufferSecurityCheck.cpp:
```diff
@@ -0,0 +1,298 @@
+
+  // Emit the CMP instruction to compare stack cookie with global cookie
+  BuildMI(*CurMBB, InsertPt, DebugLoc(), TII->get(AArch64::SUBSXrr))
+      .addReg(DiscardReg,
```
Can you use XZR as the destination register?

efriedma-quic commented on llvm/lib/Target/AArch64/AArch64WinFixupBufferSecurityCheck.cpp:
```diff
@@ -0,0 +1,298 @@
+  // argument passing
+  ++DIt;
+  if (DIt == MBB->rend() || DIt->getOpcode() != AArch64::COPY) {
+    return false;
```
Also check that the COPY destination is x0, and the source is the ldr.

efriedma-quic commented on llvm/lib/Target/AArch64/AArch64WinFixupBufferSecurityCheck.cpp:
```diff
@@ -0,0 +1,298 @@
+        auto MO = MI.getOperand(0);
+        if (MO.isGlobal()) {
+          auto Callee = dyn_cast<Function>(MO.getGlobal());
+          if (Callee && Callee->getName() == "__security_check_cookie") {
```
getSecurityCheckCookieName(), to handle arm64ec correctly.

efriedma-quic commented on llvm/lib/Target/AArch64/AArch64TargetMachine.cpp:
```diff
@@ -812,6 +812,7 @@ void AArch64PassConfig::addPreRegAlloc() {
   }
   if (TM->getOptLevel() != CodeGenOptLevel::None && EnableMachinePipeliner)
     addPass(&MachinePipelinerID);
+  addPass(createAArch64WinFixupBufferSecurityCheckPass());
```
Should we check `TM->getOptLevel()`?  Maybe also move the `isWindowsMSVCEnvironment()` check here.

efriedma-quic commented on llvm/lib/Target/AArch64/AArch64WinFixupBufferSecurityCheck.cpp:
```diff
@@ -0,0 +1,298 @@
+  MachineInstr *LoadStackGuard = nullptr;
+
+  for (auto &MBB : MF) {
+    for (auto &MI : MBB) {
```
Scanning the whole function like this is sort of unfortunate.    Not sure what the alternative looks like, though; it doesn't naturally fit into any existing pass.

efriedma-quic replied:
Actually, just thought of an alternative: use a pseudo-instruction, and expand it in AArch64TargetLowering::EmitInstrWithCustomInserter.

tamaspetz replied:
@omjavaid As an alternative solution, would it not make a difference if `AArch64TargetLowering::getSSPStackGuardCheck()` returned a `nullptr` instead? Would it not effectively inline the checks?

omjavaid replied:
Thanks! sorry it took a while @tamaspetz I think you have given a solid suggestion here. I am writing a solution around it and will post it as a separate PR.
